src/engine/renderer/tr_shader.cpp

```diff
@@ -5572,6 +5572,88 @@ bool CheckShaderNameLength( const char* func_err, const char* name, const char*
 	return true;
 }
 
+static bool CheckStage( shaderStage_t *pStage )
+{
+	struct stageCheck_t {
+		bool expected;
+		bool check;
+		bool fallback;
+		std::string name;
+	};
+
+	static const std::unordered_map<stageType_t, stageCheck_t> stageTypeCheck = {
```

Comment:
The table of boolean flags seems kinda overkill. Why not something like
```
switch( stageType ) {
  case stageType_t::ST_GLOWMAP:
  case stageType_t::ST_DIFFUSEMAP:
  case stageType_t::ST_COLORMAP:
    if (!pStage->bundle[0].image[0]) {
     Log::Warn("%s stage missing image", stageTypeNames[stageType]);
```

That way you can deduplicate the code but you only have to put the stage names in a table.

Reply:
I prefer the table and it's on purpose: I want that when a line is added for a stage type to be documented in the table, the line cannot be added without documenting every bits about that stage type. It's written this way on purpose of relying on the C++ strong typing to yell at the developer “you haven't told me everything about it!”…